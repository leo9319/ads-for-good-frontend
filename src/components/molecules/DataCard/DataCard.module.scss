@import '../../../styles/colours';

@mixin apply-size($fontSize, $fontWeight, $lineHeight, $lineClamp: 1) {
  font-size: var($fontSize);
  font-weight: $fontWeight;
  line-height: var($lineHeight);
  line-clamp: $lineClamp;
  -webkit-line-clamp: $lineClamp;
}

@mixin apply-line-clamp {
  display: -webkit-box;
  overflow: hidden;
  align-self: stretch;
  text-overflow: ellipsis;
  -webkit-box-orient: vertical;
}

.dataCardContainer {
  display: flex;
  width: 160px;
  height: 213px;
  flex-direction: column;
  align-items: flex-start;
  padding: var(--spacing-md);
  justify-content: space-between;
  background-color: $color-white;
  box-shadow: 0 1px 3px 0 $color-box-shadow;
  font-family: var(--typography-default-font-family);

  --card-border-width: 0px;
  --card-border-radius: var(--spacing-md);
  --card-background-color: unset;

  &.surface {
    border: 1px solid $color-neutral-100;
  }

  .headerContainer {
    width: 100%;
    display: grid;
    gap: var(--spacing-xs);

    .header {
      display: grid;
      gap: var(--spacing-2xs);

      .heading {
        @include apply-size(
          --typography-font-size-5,
          700,
          --typography-line-height-5
        );
        @include apply-line-clamp();

        color: $color-neutral-950;
      }

      .subHeading {
        @include apply-size(
          --typography-font-size-2,
          600,
          --typography-line-height-2,
          2
        );
        @include apply-line-clamp();

        color: $color-neutral-950;
      }
    }

    .content {
      @include apply-size(
        --typography-font-size-1,
        400,
        --typography-line-height-1,
        5
      );
      @include apply-line-clamp();

      color: $color-neutral-600;
    }
  }

  .footer {
    width: 100%;
  }

  &.size {
    &-2 {
      width: 248px;
      height: 331px;
      padding: var(--spacing-2xl);
      --card-border-radius: var(--spacing-xl);

      .headerContainer {
        .header {
          .heading {
            @include apply-size(
              --typography-font-size-8,
              600,
              --typography-line-height-8
            );
          }

          .subHeading {
            @include apply-size(
              --typography-font-size-4,
              600,
              --typography-line-height-4,
              2
            );
          }
        }

        .content {
          @include apply-size(
            --typography-font-size-3,
            400,
            --typography-line-height-3,
            7
          );
        }
      }
    }

    &-3 {
      width: 348px;
      height: 464px;
      padding: var(--spacing-2xl);
      --card-border-radius: var(--spacing-xl);

      .headerContainer {
        gap: var(--spacing-md);

        .header {
          .heading {
            @include apply-size(
              --typography-font-size-10,
              600,
              --typography-line-height-10
            );
          }

          .subHeading {
            @include apply-size(
              --typography-font-size-4,
              600,
              --typography-line-height-4,
              2
            );
          }
        }

        .content {
          @include apply-size(
            --typography-font-size-4,
            400,
            --typography-line-height-4,
            10
          );
        }
      }
    }
  }
}
