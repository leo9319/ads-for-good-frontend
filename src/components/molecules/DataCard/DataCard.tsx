import React, { ReactElement } from 'react';
import Box from '@radix-styles/atoms/Box';
import Card from '@radix-styles/atoms/Card';
import Tag from '@radix-styles/atoms/Tag';
import Text from '@radix-styles/atoms/Text';
import { classNames } from '@utils/common/classNames';
import educationLogo from '@assets/icons/emblemLogo/Education.svg';
import foodLogo from '@assets/icons/emblemLogo/Food.svg';
import healthLogo from '@assets/icons/emblemLogo/Health.svg';
import waterLogo from '@assets/icons/emblemLogo/Water.svg';
import markLogo from '@assets/icons/emblemLogo/Mark.svg';
import childProtection from '@assets/icons/emblemLogo/Child_Protection.svg';

import style from './DataCard.module.scss';

/**
 * @interface
 * The Data Card props type of {@link DataCard}
 */
export interface DataCardProps {
  /**
   * Determines the size of the Data Card
   * @defaultValue `1`
   * @property
   */
  size?: '1' | '2' | '3';
  /**
   * Controls the visual style of the Data Card
   * @defaultValue `default`
   * @property
   */
  variant?: 'default' | 'surface';
  /**
   * Text displayed next to the Logo
   * @property
   */
  logoText: string;
  /**
   * Name of the Logo to be displayed
   * @property
   */
  logoName: string;
  /**
   * Main title or heading of the Card
   * @property
   */
  heading: string;
  /**
   * Subtitle or secondary heading of the Card
   * @property
   */
  subHeading: string;
  /**
   * A brief summary or description of the Card's content
   * @property
   */
  content: string;
  /**
   * Additional class names for customizing styles
   * @property
   */
  containerClass?: string;
}

/**
 * @interface
 * The Logo object types
 */
interface ILogoObj {
  [key: string]: string;
}

/**
 * The Data Card is a structured component that organizes related content and actions.
 * It features a heading, subheading, and content, while also supporting customizable
 * sizes, visual variants, and a logo.
 *
 * @group Molecules
 * @category Component
 *
 * @param {DataCardProps} props - The Data Card props
 * @returns {ReactElement} - Data Card Component
 *
 */
export const DataCard = ({
  heading,
  subHeading,
  content,
  logoName,
  logoText,
  size = '1',
  variant = 'default',
  containerClass = '',
}: DataCardProps): ReactElement => {
  // Temporary object to obtain the path of the emblem logo
  const logoObj: ILogoObj = {
    education: educationLogo,
    food: foodLogo,
    health: healthLogo,
    water: waterLogo,
    mark: markLogo,
    child: childProtection,
  };

  // Function to add the prefix with card size
  const getSize = (size: string): string => `size-${size}`;

  const logoPath: string = logoObj[logoName];

  const newClass = classNames(
    style.dataCardContainer,
    containerClass,
    style[getSize(size)],
    style[variant]
  );

  return (
    <Card className={newClass}>
      <Box className={style.headerContainer}>
        <Box className={style.header}>
          <Text as="p" className={style.heading}>
            {heading}
          </Text>
          <Text as="p" className={style.subHeading}>
            {subHeading}
          </Text>
        </Box>
        <Text as="p" className={style.content}>
          {content}
        </Text>
      </Box>
      <Box className={style.footer}>
        <Tag showAvatar mode="default" text={logoText} iconUrl={logoPath} />
      </Box>
    </Card>
  );
};

export default DataCard;
