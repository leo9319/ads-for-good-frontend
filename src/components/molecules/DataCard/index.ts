export type { DataCardProps } from './DataCard';
export { default } from './DataCard';
export { DataCard } from './DataCard';
