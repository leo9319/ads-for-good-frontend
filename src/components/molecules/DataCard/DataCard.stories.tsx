import React from 'react';
import DataCard from './DataCard';
import type { Meta, StoryObj } from '@storybook/react';
import { Theme } from '@radix-ui/themes';

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: 'Library/Molecules/DataCard',
  component: DataCard,
  decorators: [
    Story => (
      <Theme>
        <Story />
      </Theme>
    ),
  ],
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: 'centered',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    heading: { control: 'text' },
    subHeading: { control: 'text' },
    content: { control: 'text' },
    size: { control: 'select', options: ['1', '2', '3'] },
    variant: { control: 'select', options: ['default', 'surface'] },
    logoText: { control: 'text' },
    logoName: {
      control: 'select',
      options: ['education', 'food', 'health', 'water', 'mark', 'child'],
    },
    containerClass: { control: 'text' },
  },
  // more on args: https://storybook.js.org/docs/writing-stories/args#args-object
  args: {
    heading: '4M+',
    subHeading: 'People served',
    content:
      'Our livelihood improvement programs enable millions globally to reach better outcomes',
    size: '1',
    logoName: 'education',
  },
} satisfies Meta<typeof DataCard>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args

export const Default: Story = {
  args: {
    variant: 'default',
    logoText: 'Water',
    logoName: 'water',
  },
};

export const Surface: Story = {
  args: {
    variant: 'surface',
    logoText: 'Food',
    logoName: 'food',
  },
};
