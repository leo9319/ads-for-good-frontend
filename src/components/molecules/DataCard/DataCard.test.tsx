import React from 'react';
import { render } from '@testing-library/react';
import DataCard from './DataCard';

describe('Data Card', () => {
  it('Should Default Variant render successfully', () => {
    const { baseElement } = render(
      <DataCard
        content="Our livelihood improvement programs enable millions globally to reach better outcomes"
        heading="4M+"
        logoName="water"
        logoText="Water"
        size="1"
        subHeading="People served"
        variant="default"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Should Surface Variant render successfully', () => {
    const { baseElement } = render(
      <DataCard
        content="Our livelihood improvement programs enable millions globally to reach better outcomes"
        heading="4M+"
        logoName="food"
        logoText="Food"
        size="1"
        subHeading="People served"
        variant="surface"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
