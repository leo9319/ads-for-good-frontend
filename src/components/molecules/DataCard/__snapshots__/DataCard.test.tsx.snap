// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Card Should Default Variant render successfully 1`] = `
<body>
  <div>
    <div
      class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface dataCardContainer size-1 default"
    >
      <div
        class="rt-Box headerContainer"
      >
        <div
          class="rt-Box header"
        >
          <p
            class="rt-Text mocked-class-name typography-font-family-default heading"
          >
            4M+
          </p>
          <p
            class="rt-Text mocked-class-name typography-font-family-default subHeading"
          >
            People served
          </p>
        </div>
        <p
          class="rt-Text mocked-class-name typography-font-family-default content"
        >
          Our livelihood improvement programs enable millions globally to reach better outcomes
        </p>
      </div>
      <div
        class="rt-Box footer"
      >
        <div
          class="tagComp"
          style="pointer-events: none;"
        >
          <span
            class="rt-reset rt-Badge rt-r-size-1 rt-variant-soft default"
            data-accent-color=""
            style="padding: 4px 25px; border-radius: 20px; cursor: unset;"
            tabindex="0"
          >
            <span
              class="rt-reset rt-AvatarRoot rt-r-size-1 rt-variant-soft"
              data-accent-color=""
              data-radius="full"
              style="margin-left: -19px;"
            >
              <span
                class="rt-AvatarFallback"
              />
            </span>
            Water
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Data Card Should Surface Variant render successfully 1`] = `
<body>
  <div>
    <div
      class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface dataCardContainer size-1 surface"
    >
      <div
        class="rt-Box headerContainer"
      >
        <div
          class="rt-Box header"
        >
          <p
            class="rt-Text mocked-class-name typography-font-family-default heading"
          >
            4M+
          </p>
          <p
            class="rt-Text mocked-class-name typography-font-family-default subHeading"
          >
            People served
          </p>
        </div>
        <p
          class="rt-Text mocked-class-name typography-font-family-default content"
        >
          Our livelihood improvement programs enable millions globally to reach better outcomes
        </p>
      </div>
      <div
        class="rt-Box footer"
      >
        <div
          class="tagComp"
          style="pointer-events: none;"
        >
          <span
            class="rt-reset rt-Badge rt-r-size-1 rt-variant-soft default"
            data-accent-color=""
            style="padding: 4px 25px; border-radius: 20px; cursor: unset;"
            tabindex="0"
          >
            <span
              class="rt-reset rt-AvatarRoot rt-r-size-1 rt-variant-soft"
              data-accent-color=""
              data-radius="full"
              style="margin-left: -19px;"
            >
              <span
                class="rt-AvatarFallback"
              />
            </span>
            Food
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
